File: 0215.Kth_largest_element_in_an_array/solution.py
from random import random
from typing import List


class Solution:
    def pivot(self, nums, left, right):
        v = random.randint(left, right)
        p = nums[v]
        j = left
        for i in range(left + 1, right + 1):
            if nums[i] < p:
                j += 1
                nums[i], nums[j] = nums[j], nums[i]
        nums[v], nums[j] = nums[j], nums[v]
        return j

    def find_Kth_largest(self, nums: List[int], k: int) -> int:
        """
        第k大值
        1.通过快排倒序排好数值,取第k-1个值, 时间复杂度为排序的复杂度O(nlogn)
        2.调整快排, 根据pivot, 分为左右两边，只对影响到len(nums)-k值的区间进行排序, 时间复杂度为排序的复杂度O(n)
        """
        length = len(nums)
        index = length - k
        left, right = 0, length - 1
        while True:
            p = self.pivot(nums, left, right)
            if p > index:
                right -= 1
            elif p < index:
                left += 1
            else:
                return nums[p]

File: 0215.Kth_largest_element_in_an_array/solution.go
package leetcode0215


func findKthLargest(nums []int, k int) int {
	length := len(nums)
	left, right := 0, length - 1
	index := length - k
	for {
		p := partition(nums, left, right)
		if index > p{
			left = p + 1
		}else if index < p{
			right = p - 1
		}else{
			return nums[p]
		}
	}
}

func partition(nums []int, left, right int) int {
	// left,mid,right取中值
	mid := left + (right - left) >> 1
	if nums[left] < nums[right]{
		nums[left], nums[right] = nums[right], nums[left]
	}
	if nums[right] < nums[mid]{
		nums[mid], nums[right] = nums[right], nums[mid]
	}
	if nums[left] < nums[mid]{
		nums[mid], nums[left] = nums[left], nums[mid]
	}

	p := nums[left]
	i := left
	for j:=left+1;j<=right;j++{
		if nums[j] < p{
			i++
			nums[j], nums[i] = nums[i], nums[j]
		}
	}
	nums[left], nums[i] = nums[i], nums[left]
	return i
}
